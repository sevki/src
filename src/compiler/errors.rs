use std::ops::Range;

use crate::lexer::Coord;

pub struct Errors<'a>(Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>);

impl<'a> From<Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>>
    for Errors<'a>
{
    fn from(
        value: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>,
    ) -> Self {
        Self(value)
    }
}

impl<'a> IntoIterator for Errors<'a> {
    type Item = Range<Coord>;

    type IntoIter = <Vec<std::ops::Range<Coord>> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0
            .into_iter()
            .map(|error| match error.error {
                lalrpop_util::ParseError::InvalidToken { location } => location..location,
                lalrpop_util::ParseError::UnrecognizedEof {
                    location,
                    expected: _,
                } => location..location,
                lalrpop_util::ParseError::UnrecognizedToken { token, expected: _ } => {
                    token.0..token.2
                }
                lalrpop_util::ParseError::ExtraToken { token } => token.0..token.2,
                lalrpop_util::ParseError::User { error: _ } => todo!(),
            })
            .collect::<Vec<_>>()
            .into_iter()
    }
}

fn handle_errors(
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token, &str>>,
    src: &str,
) -> String {
    let mut pretty = String::new();
    let mut last_end = Coord::default();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof {
                location: _,
                expected: _,
            } => todo!(),
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
                last_end = end;
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
