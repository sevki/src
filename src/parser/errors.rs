use crate::lexer::Coord;

pub fn pretty_errors<'input>(
    src: &'input str,
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'_>, &str>>,
) -> String {
    let mut pretty = String::new();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof { location, expected } => {
                let start = location;
                let end = location;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                pretty.push_str(&format!(
                    "error: unexpected end of file, expected one of {:?}\n",
                    expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                // expected is double quoted, so we need to remove the quotes
                // which ends up printing like   "\"(\""
                let expected: Vec<&str> = expected.iter().map(|x| &x[1..x.len() - 1]).collect();
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
