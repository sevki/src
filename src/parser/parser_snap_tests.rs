
use crate::parser::errors::pretty_errors;
use insta::assert_snapshot;


#[cfg(test)]
#[okstd::test]
fn test_empty_parser() {
    let input = " ";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()), @r###"
    Module(
        [],
    )
    "###);
}

#[okstd::test]
fn test_fn_call_parser_with_multiple_args_and_strings() {
    let input = "fn some()[] {let a = some_fnExpr(1, \"2\", 3)}";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_fn_def_parser() {
    let input = r#"fn call(a:b, b:c) [throws, awaits, execs] {
        call(1+1)
        let a = 1
    } when throws {
        raise(1)
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_effect() {
    let input = r#"effect VM: async + throws + execs {
        catch() []
        await<T>(f: Future<T>) [] -> T
        exec(arg0: string, args: stringvec) []
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_struct_parser() {
    let input = r#"struct VM {
        a: string
        b: string
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_enum_parser() {
    let input = r#"use { exec } from host

    effect Make: async + throws + execs + reads + writes {
        catch() [throws]
        await<T>(f: Future<T>) [async, throws] -> T
        exec(arg0: string, args: stringvec) [Make] -> i32
        read(name: string) [reads] -> string
        write(name: string, value: string) [writes]
    }
    
    struct Local {
        host: host
    }
    
    impl Make for Local {
        fn catch(self) [throws] {
        } 
        fn await<T>(f: Future<T>) [async, trhows] -> T {
            yield()
        }
        fn exec(self, arg0: string, args: vec<string>) [Vm] -> i32 {
            self.host.read("jobserver")
            if self.host.exec(arg0, args) {
                raise(1)
            }
        }
    }"#;
    let mut errors: Vec<lalrpop_util::ErrorRecovery<usize, crate::lexer::Token, &str>> = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}
