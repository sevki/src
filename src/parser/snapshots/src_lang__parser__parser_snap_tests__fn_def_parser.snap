---
source: src/parser/parser_snap_tests.rs
expression: "format!(\"{:#?}\", t.unwrap())"
---
Module(
    [
        Spanned(
            Location {
                offset: 0,
                line: 0,
                col: 0,
            },
            FnDef(
                FnDef(
                    Spanned(
                        Location {
                            offset: 0,
                            line: 0,
                            col: 0,
                        },
                        KeywordAndVisibility(
                            Spanned(
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                                Fn,
                                Location {
                                    offset: 2,
                                    line: 0,
                                    col: 2,
                                },
                            ),
                            Spanned(
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                                Private,
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                            ),
                        ),
                        Location {
                            offset: 2,
                            line: 0,
                            col: 2,
                        },
                    ),
                    Spanned(
                        Location {
                            offset: 3,
                            line: 0,
                            col: 3,
                        },
                        Prototype {
                            name: Spanned(
                                Location {
                                    offset: 3,
                                    line: 0,
                                    col: 3,
                                },
                                Ident(
                                    "call",
                                    None,
                                ),
                                Location {
                                    offset: 7,
                                    line: 0,
                                    col: 7,
                                },
                            ),
                            args: [
                                Spanned(
                                    Location {
                                        offset: 8,
                                        line: 0,
                                        col: 8,
                                    },
                                    Field(
                                        Spanned(
                                            Location {
                                                offset: 8,
                                                line: 0,
                                                col: 8,
                                            },
                                            FieldDef(
                                                Spanned(
                                                    Location {
                                                        offset: 8,
                                                        line: 0,
                                                        col: 8,
                                                    },
                                                    Private,
                                                    Location {
                                                        offset: 8,
                                                        line: 0,
                                                        col: 8,
                                                    },
                                                ),
                                                Spanned(
                                                    Location {
                                                        offset: 8,
                                                        line: 0,
                                                        col: 8,
                                                    },
                                                    Ident(
                                                        "a",
                                                        None,
                                                    ),
                                                    Location {
                                                        offset: 9,
                                                        line: 0,
                                                        col: 9,
                                                    },
                                                ),
                                                Spanned(
                                                    Location {
                                                        offset: 10,
                                                        line: 0,
                                                        col: 10,
                                                    },
                                                    Ident(
                                                        "b",
                                                        None,
                                                    ),
                                                    Location {
                                                        offset: 11,
                                                        line: 0,
                                                        col: 11,
                                                    },
                                                ),
                                            ),
                                            Location {
                                                offset: 11,
                                                line: 0,
                                                col: 11,
                                            },
                                        ),
                                    ),
                                    Location {
                                        offset: 11,
                                        line: 0,
                                        col: 11,
                                    },
                                ),
                                Spanned(
                                    Location {
                                        offset: 13,
                                        line: 0,
                                        col: 13,
                                    },
                                    Field(
                                        Spanned(
                                            Location {
                                                offset: 13,
                                                line: 0,
                                                col: 13,
                                            },
                                            FieldDef(
                                                Spanned(
                                                    Location {
                                                        offset: 13,
                                                        line: 0,
                                                        col: 13,
                                                    },
                                                    Private,
                                                    Location {
                                                        offset: 13,
                                                        line: 0,
                                                        col: 13,
                                                    },
                                                ),
                                                Spanned(
                                                    Location {
                                                        offset: 13,
                                                        line: 0,
                                                        col: 13,
                                                    },
                                                    Ident(
                                                        "b",
                                                        None,
                                                    ),
                                                    Location {
                                                        offset: 14,
                                                        line: 0,
                                                        col: 14,
                                                    },
                                                ),
                                                Spanned(
                                                    Location {
                                                        offset: 15,
                                                        line: 0,
                                                        col: 15,
                                                    },
                                                    Ident(
                                                        "c",
                                                        None,
                                                    ),
                                                    Location {
                                                        offset: 16,
                                                        line: 0,
                                                        col: 16,
                                                    },
                                                ),
                                            ),
                                            Location {
                                                offset: 16,
                                                line: 0,
                                                col: 16,
                                            },
                                        ),
                                    ),
                                    Location {
                                        offset: 16,
                                        line: 0,
                                        col: 16,
                                    },
                                ),
                            ],
                            ret: None,
                            effects: [
                                Spanned(
                                    Location {
                                        offset: 19,
                                        line: 0,
                                        col: 19,
                                    },
                                    Ident(
                                        "throws",
                                        None,
                                    ),
                                    Location {
                                        offset: 25,
                                        line: 0,
                                        col: 25,
                                    },
                                ),
                                Spanned(
                                    Location {
                                        offset: 27,
                                        line: 0,
                                        col: 27,
                                    },
                                    Ident(
                                        "awaits",
                                        None,
                                    ),
                                    Location {
                                        offset: 33,
                                        line: 0,
                                        col: 33,
                                    },
                                ),
                                Spanned(
                                    Location {
                                        offset: 35,
                                        line: 0,
                                        col: 35,
                                    },
                                    Ident(
                                        "execs",
                                        None,
                                    ),
                                    Location {
                                        offset: 40,
                                        line: 0,
                                        col: 40,
                                    },
                                ),
                            ],
                        },
                        Location {
                            offset: 41,
                            line: 0,
                            col: 41,
                        },
                    ),
                    Block(
                        [
                            Spanned(
                                Location {
                                    offset: 48,
                                    line: 1,
                                    col: 4,
                                },
                                FnCall(
                                    FnCall(
                                        Spanned(
                                            Location {
                                                offset: 48,
                                                line: 1,
                                                col: 4,
                                            },
                                            Ident(
                                                "call",
                                                None,
                                            ),
                                            Location {
                                                offset: 52,
                                                line: 1,
                                                col: 8,
                                            },
                                        ),
                                        [
                                            Spanned(
                                                Location {
                                                    offset: 54,
                                                    line: 1,
                                                    col: 10,
                                                },
                                                BinaryExpression(
                                                    BinaryOperation {
                                                        lhs: Spanned(
                                                            Location {
                                                                offset: 53,
                                                                line: 1,
                                                                col: 9,
                                                            },
                                                            Integer(
                                                                1,
                                                            ),
                                                            Location {
                                                                offset: 54,
                                                                line: 1,
                                                                col: 10,
                                                            },
                                                        ),
                                                        op: Add,
                                                        rhs: Spanned(
                                                            Location {
                                                                offset: 55,
                                                                line: 1,
                                                                col: 11,
                                                            },
                                                            Integer(
                                                                1,
                                                            ),
                                                            Location {
                                                                offset: 56,
                                                                line: 1,
                                                                col: 12,
                                                            },
                                                        ),
                                                    },
                                                ),
                                                Location {
                                                    offset: 55,
                                                    line: 1,
                                                    col: 11,
                                                },
                                            ),
                                        ],
                                    ),
                                ),
                                Location {
                                    offset: 57,
                                    line: 1,
                                    col: 13,
                                },
                            ),
                            Spanned(
                                Location {
                                    offset: 62,
                                    line: 2,
                                    col: 4,
                                },
                                Binding(
                                    Binding(
                                        Spanned(
                                            Location {
                                                offset: 66,
                                                line: 2,
                                                col: 8,
                                            },
                                            Ident(
                                                "a",
                                                None,
                                            ),
                                            Location {
                                                offset: 67,
                                                line: 2,
                                                col: 9,
                                            },
                                        ),
                                        Spanned(
                                            Location {
                                                offset: 70,
                                                line: 2,
                                                col: 12,
                                            },
                                            Integer(
                                                1,
                                            ),
                                            Location {
                                                offset: 71,
                                                line: 2,
                                                col: 13,
                                            },
                                        ),
                                    ),
                                ),
                                Location {
                                    offset: 71,
                                    line: 2,
                                    col: 13,
                                },
                            ),
                        ],
                    ),
                ),
            ),
            Location {
                offset: 73,
                line: 3,
                col: 1,
            },
        ),
    ],
)
